pkg/config/config.go

```diff
@@ -0,0 +1,133 @@
+	kubeMajorVer, err := strconv.Atoi(kubeVer[0])
+	if err != nil {
+		glog.Fatalf("Incorrect content in --kubeVersion %s", config.KubeVersion)
+	}
+	kubeMinorVer, err := strconv.Atoi(kubeVer[1])
+	if err != nil {
+		glog.Fatalf("Incorrect content in --kubeVersion %s", config.KubeVersion)
+	}
+	return kubeMajorVer, kubeMinorVer
+}
+
+// GetStatsServerAddress returns the StatsServerAddress from the config
+// TODO(shiv): We need to have separate port and IP for stats server too like firmament address amd port.
+// This separation is required when passing address as command line flags in deployment yaml,
+// the IP:PORT in the yaml throws an error.
+func GetStatsServerAddress() string {
+	return config.StatsServerAddress
+}
+
+// GetSchedulingInterval return the scheduling interval from config
+func GetSchedulingInterval() int {
+	return config.SchedulingInterval
+}
+
+// GetConfigPath returns the config path from  config
+func GetConfigPath() string {
+	return config.ConfigPath
+}
+
+// ReadFromConfigFile to read from yaml,json,toml etc poseidonConfig file
+// Note:
+//  The poseidonConfig values will be overwritten if flag for the same key are present
+func ReadFromConfigFile() {
+	viper.AddConfigPath(".")
+	viper.AddConfigPath(config.ConfigPath)
+	viper.SetConfigName("poseidon_config")
+	err := viper.ReadInConfig()
+	if err != nil {
+		glog.Warning(err, "unable to read poseidon_config, using command flags/default values")
+		return
+	}
+	err = viper.Unmarshal(&config)
+	if err != nil {
+		glog.Fatal("unmarshal poseidon_config file failed", err)
+	}
+	glog.Info("ReadFromConfigFile", config)
+}
+
+// ReadFromCommandLineFlags reads command line flags and these will override poseidonConfig file flags.
+func ReadFromCommandLineFlags() {
+	pflag.StringVar(&config.SchedulerName, "schedulerName", "poseidon", "The scheduler name with which pods are labeled")
+	pflag.StringVar(&config.FirmamentAddress, "firmamentAddress", "firmament-service.kube-system", "Firmament scheduler service port")
+	pflag.StringVar(&config.FirmamentPort, "firmamentPort", "9090", "Firmament scheduler service port")
+	pflag.StringVar(&config.KubeConfig, "kubeConfig", "kubeconfig.cfg", "Path to the kubeconfig file")
+	pflag.StringVar(&config.KubeVersion, "kubeVersion", "1.6", "Kubernetes version")
+	pflag.StringVar(&config.StatsServerAddress, "statsServerAddress", "0.0.0.0:9091", "Address on which the stats server listens")
+	pflag.IntVar(&config.SchedulingInterval, "schedulingInterval", 10, "Time between scheduler runs (in seconds)")
+	pflag.StringVar(&config.ConfigPath, "configPath", ".",
+		"The path to the config file (i.e poseidon_cfg) without filename or extension, supported extensions/formats are Yaml, Json")
+
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+	pflag.Parse()
+	viper.BindPFlags(pflag.CommandLine)
+	glog.Info("ReadFromCommandLineFlags", config)
+}
+
+func init() {
+	ReadFromCommandLineFlags()
+	ReadFromConfigFile()
```

Comment:
Does it mean config file will overwrite command line flags?

Plus, I would suggest do not put `ReadFromCommandLineFlags` and `ReadFromConfigFile` in the `init()` in case we will call them in other places, e.g. unit tests.

Reply:
No, if the config file has the same fields set by the command line flags, it will use command line flags only, and ignore config file flags. viper package does this job.

In case of default values set by the command line flags, if the config file has same fields specified, config file flags will take precedence over the default value from the command line flag.

Reply:
Okay, thanks for your clarification.